File: pset1/credit.c
#include <stdio.h>
#include <cs50.h>

int main(void)
{
    long numberType = get_long("Number: ");
    long number = numberType;
    long numberHead1;
    long numberHead2;
    long cardValidateHead;
    int count = 0;
    int checksum = 0;

    while (number != 0)
    {
        if (count % 2 != 0)
        {
            checksum = checksum + (number % 10) * 2 / 10 + (number % 10) * 2 % 10;
        }
        else
        {
            checksum = checksum + number % 10;  
        }

        //record the first and second number 
        numberHead2 = numberHead1;
        numberHead1 = number % 10;
            
        number = number / 10;
        count++;
        // printf("numberHead1: %ld, numberHead2: %ld\n", numberHead1, numberHead2);
    }      

    checksum = checksum % 10;
    cardValidateHead = numberHead1 * 10 + numberHead2;
    // printf("%d\n", checksum);
    // printf("%d\n", count);s
    if (checksum == 0)
    {
        if (count == 15 && ((cardValidateHead == 34) || (cardValidateHead == 37))) 
        {
            printf("AMEX\n");
        }
        else if (count == 16 && ((cardValidateHead >= 51) && (cardValidateHead <= 55)))
        {
            printf("MASTERCARD\n");  
        }
        else if ((numberHead1 == 4) && (count == 13 || count == 16))
        {
            printf("VISA\n");  
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }        
}

File: pset6/credit/credit.py
from cs50 import get_int, get_string

num = get_string("Number: ")

evenNum = 0
oddNum = 0
length = len(num)
number = num[::-1]
# print(number)
for even in range(1, length, 2):
    if int(number[even]) * 2 >= 10:
        # print(number[even])
        evenCount = int(number[even]) * 2 // 10 + int(number[even]) * 2 % 10
        # print(evenCount)
        evenNum = evenNum + evenCount
    else:
        # print(int(number[even]) * 2)
        evenNum = evenNum + int(number[even]) * 2
# print(evenNum)
    
for odd in range(0, length, 2):
    oddNum = oddNum + int(number[odd])
# print(oddNum)

cardValidateHead = 10 * int(num[0]) + int(num[1])

checksum = (evenNum + oddNum) % 10

if checksum == 0:
    if length == 15 and ((cardValidateHead == 34) or (cardValidateHead == 37)): 
        print("AMEX")
    elif length == 16 and ((cardValidateHead >= 51) and (cardValidateHead <= 55)):
        print("MASTERCARD") 
    elif int(num[0]) == 4 and (length == 13 or length == 16):
        print("VISA")  
    else:
        print("INVALID")
else:
    print("INVALID")
